File: leetcode/704.二分查找.py
#
# @lc app=leetcode.cn id=704 lang=python3
#
# [704] 二分查找
#

# @lc code=start
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        low, high = 0, len(nums) - 1
        while low <= high:
            mid = (high - low) // 2 + low  #防止溢出，当low和high很大时候，相加会溢出，相减就不会。
            num = nums[mid]
            if num == target:
                return mid
            elif num > target:
                high = mid - 1
            else:
                low = mid + 1
        return -1

# @lc code=end

File: leetcode/704.二分查找.go
/*
 * @lc app=leetcode.cn id=704 lang=golang
 *
 * [704] 二分查找
 */
package leetcode
// @lc code=start
func search(nums []int, target int) int {
    low, high := 0, len(nums) -1
    for low <= high{
        mid := (high - low)/2 + low
        num := nums[mid]
        if num == target{
            return mid
        }else if num > target{
            high = mid - 1
        }else{
            low = mid + 1
        }
    }
    return -1
}
// @lc code=end

